### ISSUE TITLE
Suitesparse compiled without OpenMP (and without _POSIX_C_SOURCE enabled)

### ISSUE THREAD

**Issue by freeseek** [Homebrew/homebrew-core#129011]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
brew suitesparse
```

### Verification

- [X] My "`brew doctor` output" says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.
- [X] I searched for recent similar issues at https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue and found no duplicates.

### What were you trying to do (and why)?

I compiled a software based on CHOLMOD (a library included with suitesparse) and when running the function SuiteSparse_tic() to extract the time (and some other SuiteSparse functionalities) does not work.

I believe the problem is that in SuiteSparse, the function SuiteSparse_tic() is defined as in [SuiteSparse_config.c](https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/dev/SuiteSparse_config/SuiteSparse_config.c) as:
```
#if !defined ( SUITESPARSE_TIMER_ENABLED )

    /* ---------------------------------------------------------------------- */
    /* no timer */
    /* ---------------------------------------------------------------------- */

    void SuiteSparse_tic
    (
        double tic [2]      /* output, contents undefined on input */
    )
    {
        /* no timer installed */
        tic [0] = 0 ;
        tic [1] = 0 ;
    }

#elif defined ( _OPENMP )

    /* ---------------------------------------------------------------------- */
    /* OpenMP timer */
    /* ---------------------------------------------------------------------- */

    void SuiteSparse_tic
    (
        double tic [2]      /* output, contents undefined on input */
    )
    {
        tic [0] = omp_get_wtime ( ) ;
        tic [1] = 0 ;
    }

#else 

    /* ---------------------------------------------------------------------- */
    /* POSIX timer */
    /* ---------------------------------------------------------------------- */

    #include <time.h>
    void SuiteSparse_tic
    (
        double tic [2]      /* output, contents undefined on input */
    )
    {
        /* POSIX C 1993 timer, requires -lrt */
        struct timespec t ;
        clock_gettime (CLOCK_MONOTONIC, &t) ;
        tic [0] = (double) (t.tv_sec) ;
        tic [1] = (double) (t.tv_nsec) ;
    }

#endif
```
And the macro `SUITESPARSE_TIMER_ENABLED` is defined in [SuiteSparse_config.h](https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/dev/SuiteSparse_config/SuiteSparse_config.h) as:
```
#ifndef NTIMER
    #if defined ( _OPENMP )
        #define SUITESPARSE_TIMER_ENABLED
    #elif defined ( _POSIX_C_SOURCE )
        #if _POSIX_C_SOURCE >= 199309L
        #define SUITESPARSE_TIMER_ENABLED
        #endif
    #endif
#endif
```
So that the issue arises from the fact that when compiling with clang both `_OPENMP` and `_POSIX_C_SOURCE` are not defined and so the timer is not enabled. There are two possible solutions:
- make sure SuiteSparse is compiled with OpenMP (preferred option as it would improve the efficiency of the library)
- make sure SuiteSparse is compiled with the `_POSIX_C_SOURCE` macro active (will at least recover the timer)

### What happened (include all command output)?

The SuiteSparse_tic() function returned zero

### What did you expect to happen?

The SuiteSparse_tic() function should have returned the current time

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
Create the following `main.c` file:

#include <stdio.h>
#include <SuiteSparse_config.h>

int main() {
  double toc [2] ;
  SuiteSparse_tic (toc) ;
  printf("%f\n", toc [0] + 1e-9 * toc [1]);
}
```
Then run:
```
clang main.c -lsuitesparseconfig && ./a.out
```
It will return:
```
0.000000
```
While it should return something like:
```
1682111871.575413
```
```

**Comment by SMillerDev** [comment]
What is the bug in brew exactly? This sounds like a combination of a feature request for OpenMP and an issue with upstreams detection of some features.

**Comment by freeseek** [comment]
I am not too familiar with brew (and with Mac in general), but I thought that this is related to how [suite-sparse](https://github.com/Homebrew/homebrew-core/blob/master/Formula/suite-sparse.rb) is compiled as a package. Even leaving OpenMP aside, my understanding is that the library could be compiled with this missing function if `gcc` was used instead of `clang` or if `clang` was used with the `_POSIX_C_SOURCE` macro defined. Similarly, the library could be compiled with OpenMP support (and depend on [libomp](https://formulae.brew.sh/formula/libomp)), but it is not clear to me what would need to change

**Comment by SMillerDev** [comment]
We use the system compiler so on Linux it'll use GCC and on macOS we use CLang. If that means it's missing something, that should be reported upstream since Homebrew uses the default build structure wherever it can.

**Comment by freeseek** [comment]
Okay, sorry my understanding of brew does not allow me to be more helpful, but does that mean that there is no way to write a package formula that defines `_POSIX_C_SOURCE` (to an appropriate value) when using `clang` and leave it as is when using `gcc`? If so, I will file a bug with the maintainer of `suitesparse` to see if there is a way to get it sorted upstream. As for OpenMP, the current situation is quite suboptimal because `suitesparse` is meant to be a highly performant library, which would not be the optimal case if it is not parallelized

**Comment by SMillerDev** [comment]
There must be ways to set `_POSIX_C_SOURCE`. But brew shouldn't. It is up to upstream to set that correctly, not up to the packager.

**Comment by carlocab** [comment]
Yea, it seems like we should be using OpenMP here, based on upstream's build documentation.

I'll try to fix this in #129100 by adding a dependency on `libomp`.

I'm concerned about using a different OpenMP implementation than the one used in the `openblas` dependency, however.
